services/requester/blocks_provider.go

```diff
@@ -0,0 +1,108 @@
+package requester
+
+import (
+	ethTypes "github.com/onflow/flow-evm-gateway/eth/types"
+	"github.com/onflow/flow-evm-gateway/models"
+	"github.com/onflow/flow-evm-gateway/storage"
+	"github.com/onflow/flow-go/fvm/evm/offchain/blocks"
+	evmTypes "github.com/onflow/flow-go/fvm/evm/types"
+	flowGo "github.com/onflow/flow-go/model/flow"
+	gethCommon "github.com/onflow/go-ethereum/common"
+	"github.com/onflow/go-ethereum/eth/tracers"
+)
+
+type blockSnapshot struct {
+	*BlocksProvider
+	block models.Block
+}
+
+var _ evmTypes.BlockSnapshot = (*blockSnapshot)(nil)
+
+func (bs *blockSnapshot) BlockContext() (evmTypes.BlockContext, error) {
+	blockContext, err := blocks.NewBlockContext(
+		bs.chainID,
+		bs.block.Height,
+		bs.block.Timestamp,
+		func(n uint64) gethCommon.Hash {
+			block, err := bs.blocks.GetByHeight(n)
+			if err != nil {
+				return gethCommon.Hash{}
+			}
+			blockHash, err := block.Hash()
+			if err != nil {
+				return gethCommon.Hash{}
+			}
+
+			return blockHash
+		},
+		bs.block.PrevRandao,
+		bs.tracer,
+	)
+	if err != nil {
+		return evmTypes.BlockContext{}, err
+	}
+
+	if bs.blockOverrides == nil {
+		return blockContext, nil
+	}
+
+	if bs.blockOverrides.Number != nil {
+		blockContext.BlockNumber = bs.blockOverrides.Number.ToInt().Uint64()
+	}
+
+	if bs.blockOverrides.Time != nil {
+		blockContext.BlockTimestamp = uint64(*bs.blockOverrides.Time)
+	}
+
+	if bs.blockOverrides.Random != nil {
+		blockContext.Random = *bs.blockOverrides.Random
+	}
+
+	if bs.blockOverrides.Coinbase != nil {
+		blockContext.GasFeeCollector = evmTypes.NewAddress(*bs.blockOverrides.Coinbase)
+	}
+
+	return blockContext, nil
+}
+
+type BlocksProvider struct {
+	blocks         storage.BlockIndexer
+	chainID        flowGo.ChainID
+	tracer         *tracers.Tracer
+	blockOverrides *ethTypes.BlockOverrides
+}
+
+var _ evmTypes.BlockSnapshotProvider = (*BlocksProvider)(nil)
+
+func NewBlocksProvider(
+	blocks storage.BlockIndexer,
+	chainID flowGo.ChainID,
+) *BlocksProvider {
+	return &BlocksProvider{
+		blocks:  blocks,
+		chainID: chainID,
+	}
+}
+
+func (bp *BlocksProvider) SetTracer(tracer *tracers.Tracer) {
+	bp.tracer = tracer
+}
+
+func (bp *BlocksProvider) SetBlockOverrides(blockOverrides *ethTypes.BlockOverrides) {
```

Comment:
I'm not sure this is safe. It looks like we are using the blocks provider in multiple goroutines.

I would recommend changing this to something like:
```
func (bp *BlocksProvider) WithBlockOverrides(blockOverrides *ethTypes.BlockOverrides) evmTypes.BlockSnapshotProvider {
	return &BlocksProviderWithOverrides{
		BlocksProvider: *bp,
		blockOverrides: blockOverrides,
	}
}


type BlocksProviderWithOverrides struct {
	BlocksProvider

	blockOverrides *ethTypes.BlockOverrides
}

Reply:
Nice catch :100: I have updated the API in https://github.com/onflow/flow-evm-gateway/pull/691/commits/da79cd348f51382621557902c7176bd10fe494b6 . The `requester` type doesn't need to hold a single `BlocksProvider` object, it's safe to use a new one, and inject the optional objects as needed (tracer & block overrides).